File: 2698-Find-the-Punishment-Number-of-an-Integer/ts/main.ts
function punishmentNumber(n: number): number {
    let answer = 0;

    // Loop through all numbers from 1 to n
    for (let i = 1; i <= n; i++) {
        answer += calculatePunishment(i);
    }
    return answer;
}

// Helper function to calculate the punishment for each number
function calculatePunishment(n: number): number {
    const sq = n * n;
    const sqStr = sq.toString();

    const partitions = generatePartitions(sqStr);

    // Check each partition to see if its sum matches n
    for (const partition of partitions) {
        let sum = 0;
        for (const part of partition) {
            const num = parseInt(part, 10);
            if (!isNaN(num)) {
                sum += num;
            }
        }

        // If the sum matches n, return the square
        if (sum === n) {
            return sq;
        }
    }

    return 0;
}

// Helper function to generate all partitions of a string
function generatePartitions(s: string): string[][] {
    const result: string[][] = [];
    const partition: string[] = [];
    helper(s, 0, partition, result);
    return result;
}

// Recursive helper function to backtrack and generate partitions
function helper(s: string, start: number, partition: string[], result: string[][]): void {
    if (start === s.length) {
        result.push([...partition]);
        return;
    }

    // Loop to generate all partitions
    for (let i = start + 1; i <= s.length; i++) {
        const substr = s.substring(start, i);
        partition.push(substr);
        helper(s, i, partition, result);
        partition.pop(); // Backtrack
    }
}

File: 2698-Find-the-Punishment-Number-of-an-Integer/rs/main.rs
impl Solution {
    pub fn punishment_number(n: i32) -> i32 {
        let mut answer = 0;
        for i in 1..=n {
            answer += Self::calculate_punishment(i);
        }
        answer
    }

    fn calculate_punishment(n: i32) -> i32 {
        let sq = n * n;
        let sq_str = sq.to_string();

        let partitions = Self::generate_partitions(&sq_str);

        // Check each partition and calculate its sum
        for partition in partitions {
            let mut sum = 0;
            for part in partition {
                if let Ok(num) = part.parse::<i32>() {
                    sum += num;
                }
            }

            // If the sum matches n, return the square
            if sum == n {
                return sq;
            }
        }

        0
    }

    fn generate_partitions(s: &str) -> Vec<Vec<String>> {
        let mut result = Vec::new();
        let mut partition = Vec::new();
        Self::helper(s, 0, &mut partition, &mut result);
        result
    }

    fn helper(s: &str, start: usize, partition: &mut Vec<String>, result: &mut Vec<Vec<String>>) {
        if start == s.len() {
            result.push(partition.clone());
            return;
        }

        // Loop to generate all partitions
        for i in start + 1..=s.len() {
            let substr = &s[start..i];
            partition.push(substr.to_string());
            Self::helper(s, i, partition, result);
            partition.pop(); // Backtrack
        }
    }
}
